pilot/pkg/networking/core/v1alpha3/gateway.go

```diff
@@ -185,6 +185,41 @@ func (configgen *ConfigGeneratorImpl) buildGatewayListeners(builder *ListenerBui
 	return builder
 }
 
+func buildNameToServiceMapForHTTPRoutes(push *model.PushContext,
+	virtualService config.Config) map[host.Name]*model.Service {
+
+	vs := virtualService.Spec.(*networking.VirtualService)
+	nameToServiceMap := map[host.Name]*model.Service{}
+
+	addService := func(hostname host.Name) {
+		if _, exist := nameToServiceMap[hostname]; exist {
+			return
+		}
+		// First, we obtain the service which has the same namespace as virtualService
+		if s, exist := push.ServiceIndex.HostnameAndNamespace[hostname][virtualService.Namespace]; exist {
+			nameToServiceMap[hostname] = s
+		} else {
+			// Didn't find match for our namespace, pick an arbitrary one.
+			// We just get service based on hostname directly.
+			nameToServiceMap[hostname] = push.ServiceIndex.Hostname[hostname]
```

Comment:
donot use arbitrary one, should check the svc's exportTo

Reply:
So what i mean is: maybe we could remove serviceIndex.Hostname totally, all this can be derived from `HostnameAndNamespace`

Reply:
We have a method `PushContext.ServiceForHostname`, you can use that

Reply:
> donot use arbitrary one, should check the svc's exportTo

Does your idea is that if we find no match service for our namespace, we use hostname key to iterate the `HostnameAndNamespace` map for every existing namespace to find one service exporting to our namespace? Then if there is no one service which is visible to our namespace, what should we do? Do we need the fallback policy to pick an arbitrary one? Now, I known the same fallback policy is applied in sidecar case.

Reply:
> So what i mean is: maybe we could remove serviceIndex.Hostname totally, all this can be derived from `HostnameAndNamespace`

Agree with you. I will remove the Hostname field from serviceIndex.

Reply:
> We have a method `PushContext.ServiceForHostname`, you can use that

This method `PushContext.ServiceForHostname` just pick an arbitrary one for given hostname.